server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1070,6 +1064,12 @@ private void deleteShardSnapshot(IndexId indexId, ShardId snapshotShardId, Snaps
         // finalize the snapshot and rewrite the snapshot index with the next sequential snapshot index
         finalizeShard(newSnapshotsList, fileListGeneration, blobs, "snapshot deletion [" + snapshotId + "]", shardContainer,
             snapshotShardId, snapshotId);
+
+        try {
+            indexShardSnapshotFormat.delete(shardContainer, snapshotId.getUUID());
```

Comment:
this will now throw and log an exception in case where we're deleting the last snapshot of an index, given that we delete this file already in `finalizeShard`.
Perhaps we should split the `finalizeShard` method in two, first where we write the new file generation, and then have the clean-up separate. I also wonder if the clean-up there should only be happening in case of  a delete  instead of both snapshot  and delete, similar as we do  with orphaned indices and top-level files.

Reply:
@ywelsch no we don't delete the file in `finalizeShard` yet (we don't clean up stale `snap-` references ... which is a problem in and of itself but I'd open a PR for that after this one).

As far as the cleanup only on delete, that's fine with me. Much better to align root level and shard level functionality as much as we can imo :)

Reply:
> no we don't delete the file in finalizeShard yet

we do in case of `snapshots.isEmpty`, which is what I said above.

```
if (snapshots.isEmpty()) {
    // If we deleted all snapshots, we don't need to create a new index file and simply delete all the blobs we found
    blobsToDelete = List.copyOf(blobs.keySet());
}
```

Reply:
:man_facepalming: I'm sorry , fixed now in https://github.com/elastic/elasticsearch/pull/44165/commits/47e39f6b47f0db9fd433175a5f7cfbf025f2f2b4

Reply:
> As far as the cleanup only on delete, that's fine with me. Much better to align root level and shard level functionality as much as we can imo :)

++